### ISSUE TITLE
handling CORS

### ISSUE THREAD

**Issue by gruppjo** [mwaylabs/generator-m-ionic#285]
while using chrome with --disable-web-security serves its purpose I find it a little hacky. Discussion stimulated by @landrysoules on gitter https://gitter.im/mwaylabs/generator-m-ionic?at=562f570ce16589782d92f3a5

http://blog.ionic.io/handling-cors-issues-in-ionic/

**Comment by MathiasTim** [comment]
I would also be happy if we could fix the old `--disable-web-security` habit. Using a proxy seems to be a good solution to me. 

We had the same discussion in the javaScript meeting at our company. A colleague developing for web only also said he uses a proxy script, I think [this one](https://github.com/drewzboto/grunt-connect-proxy), but I'm not quite sure.

**Comment by MathiasTim** [comment]
For all people running in this, Chrome Canaray has changed the behavior, the new command to disabled web security is: (for OSX)
`open -a Google\ Chrome\ Canary --args --disable-web-security --user-data-dir=$HOME/profile-folder-name'`

the `--user-data-dir`-part is new and required

**Comment by gruppjo** [comment]
I found this really cumbersome as well, especially since the `--user-data-dir` arg doesn't understand `~` as your home directory. So if you want to use your usual Canary Profile you can't just run `canary --disable-web-security --user-data-dir="~/Library/Application Support/Google/Chrome Canary/"`, you need to use $HOME like so `canary --disable-web-security --user-data-dir="$HOME/Library/Application Support/Google/Chrome Canary/"`. 

It took me a while to figure out where all my tabs went... Really annoying.

**Comment by gruppjo** [comment]
I put some extensive thoughts and code into this and I'm not sure how we can get around the `--disable-web-security` practice in a user-friendly fashion and in particular I'm not sure whether the generator is a good place to resolve this. So I'm dumping the result of my thoughts here to get some clarification for myself and possibly others (wink at @MathiasTim).

### TL;DR/Conclusion
From my perspective there's just too many things that could go wrong, so many edge cases and quite a big and complex amount of code to write that will be hard to maintain and include in the already existing workflow with little benefit (actually none if CORS is not an issue) and above all it'll be very confusing for everyone. 

The pain/gain-ration on this one is just not worth it, so why not just use this [CORS plugin](https://chrome.google.com/webstore/detail/allow-control-allow-origi/nlfbmbojpeacfghkpbjhddihlkkiljbi) or possibly write our own? It didn't work in all the cases for me when working on the last project but [the source](https://github.com/vitvad/Access-Control-Allow-Origin) is fairly simple and we could easily contribute. Then the CORS issue is dealt with within seconds.

We can then write a guide on how to install/configure the plugin. I would like that much better. What do you think?

### Requirements
If we implement a solution I have the following requirements:
- **Using the proxy should be transparent**. I don't want to have to fiddle around in many files, configuring URLs and what not in every project, especially not in the generated gulp files or app files. I want to leave the former intact for the sake of easy updating of the project's generator version and the latter to avoid unnecessary complexity.
- **Multiple end-points should be possible**: A very common use case is that the app needs to retrieve data from http://api1.com (resource server) and send data to http://api2.com (push server)
- Having **several watch tasks in different projects** should still be possible (proxy URLs need to be dynamic)

### CORS example use-case
**Note**: CORS is only an issue when testing in the browser AND when the given API Endpoint is improperly configured. The best would be when the API Endpoint is properly configured. I know this is often not the case and sometimes can hardly be influenced.

Let's assume we have 3 endpoints, all of them don't have CORS configured correctly:
```js
{
  "endpoint1": "http://endpoint1.com",
  "endpoint2": "http://endpoint2.com",
  "endpoint3": "http://endpoint3.com"
}
```
The proxy then needs 3 URLs to know which request to send to which endpoint:
```js
{
  "proxyURL1": "localhost:3000/proxy1", // => endpoint1
  "proxyURL2": "localhost:3000/proxy2", // => endpoint2
  "proxyURL3": "localhost:3000/proxy3"  // => endpoint3
}
```

If we try to get around CORS using the generator the following steps are needed:

1. create a **proxy that listens on multiple URL**s so it's possible to bypass CORS for multiple end-points 
  - retrieve the dynamically allocated port from `browserSync` when starting watch in order not to break the ability to have several concurrent watch tasks in different projects and have several proxies running at the same time
  - with that port create the proxyURLs and listen on them (restriction: no modules with the name of the proxy path, e.g. `proxy1`, `proxy2, `proxy3`, can be used anymore)
  - tell the proxy how to map the proxyURLs to the endpoints (directly in the gulp file, in a config file, via the command line)
2. manipulate the app code to replace endpoint1 with their according mapping
  - option 1: interceptor 
    - needs to be injected into the code, appended to an existing module (complex tasks to solve programatically: which modules exist?)
     - explicit code changes: danger of ending up in a commit
  - option 2: environments
    - addition environment needed or manipulating an existing one
    - explicit code changes: danger of ending up in a commit
  - option 3: run through all files and replace the URLs with their proxy mapping (that's what the [ionic blog](http://blog.ionic.io/handling-cors-issues-in-ionic/) suggests)
    - massive overhead and increased complexity of already complex watch cycle

**Comment by gruppjo** [comment]
For the simple proxy I created two implementations.
use: `gulp watch --no-open --proxyPath=/proxy-path --proxyMapTo=https://echo.getpostman.com`

using https://github.com/andrewrk/node-proxy-middleware (has no logging but more intuitive to use):
```js
    var url = require('url');
    var proxy = require('proxy-middleware');
    var proxyOptions = url.parse(options.proxyMapTo);
    proxyOptions.route = options.proxyPath;

    bsOptions.server.middleware = [proxy(proxyOptions)];
```

using https://github.com/chimurai/http-proxy-middleware. Has logging but wasn't able to configure the proxy replacements the way I wanted so far.
```js
    var pathRewrite = {};
    pathRewrite[options.proxyPath] = '';
    var proxyMiddleware =  require('http-proxy-middleware');
    var proxy = proxyMiddleware(options.proxyPath, {
      target: options.proxyMapTo,
      pathRewrite: pathRewrite
    });
    bsOptions.server.middleware = [proxy];
```

**Comment by gruppjo** [comment]
I wrote a guide: https://github.com/mwaylabs/generator-m-ionic/blob/dev/docs/guides/cors_proxy.md#disable-cors-in-chrome
Try out the implementation here: https://github.com/mwaylabs/generator-m-ionic-demo/tree/dev
